### PR TITLE
Adding a way to publish single nuget package when a scope is provided…

### PR DESCRIPTION
**Pull request by shahabhijeet**
1) Adding a way to publish single nuget package when a scope is provided.
2) This change will also allow publishing multiple packages from the same scope (only NetCore projects are supported as this stage). Due to restrictions on how we have enabled non-NetCore project imports and the way msbuild builds it's build graph.
3) Adding test collateral and tests to test publishing one project

### CODE REVIEW

File: build.proj [hunk 1]
```diff
@@ -131,16 +130,20 @@
     
     <!--Do not build old libraries except Authentication, which Azure PowerShell still uses-->
     <ItemGroup>
-      <NonAutoRestLibraries Include="$(LibrarySourceFolder)\Authentication\Authentication.sln"
+	<NonAutoRestLibraries Include="$(LibrarySourceFolder)\Authentication\Authentication.sln"	  
                             Condition=" '$(Scope)' == 'all' or '$(Scope)' == 'Authentication' " />
+    <!--
+		<NonAutoRestLibraries Include="$(LibrarySourceFolder)\Authentication\Authentication.sln"
+                            Condition=" ('$(Scope)' == 'all' or '$(Scope)' == 'Authentication') and Exists('$(LibrarySourceFolder)\Authentication\Authentication.sln')" />
```
Comment (hovsepm) [inline build.proj]: please uncomment or delete this line.

File: tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs [hunk 2]
```diff
@@ -43,6 +52,18 @@ public override bool Execute()
...
+            List<string> nPkgsList = null;
+
+            if (NugetPackagesToPublish != null)
+            {
+                NugetPackagesToPublish = NugetPackagesToPublish.Trim();
+
+                if (!string.IsNullOrEmpty(NugetPackagesToPublish))
+                {   
+                    nPkgsList = NugetPackagesToPublish.Split(' ').ToList<string>();
+                }
```
Comment (hovsepm) [inline tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs]: you should Split using IgnoreWhiteSpaces parameter. if there will be 2 spaces or more in the passed parameter this part may not work correct

**Comment by hovsepm** [comment]
> using System;

add license header. 

---
Refers to: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs:1 in e476727. [](commit_id = e476727f99b8f31bc6c6d3efe35a18c5fa571505, deletion_comment = False)

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CSProjTestPublish
+{
+    public class Class1
+    {
+    }
```
Comment (hovsepm) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs]: please add at least a method that returns "hello published world!" :)

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+using System.Reflection;
```
Comment (hovsepm) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Properties/AssemblyInfo.cs]: License header missing

**Comment by hovsepm** [comment]
Is this PR intended to be merged?

File: build.proj (hunk 1 above)
Comment (shahabhijeet) [inline build.proj]: done

**Comment by shahabhijeet** [comment]
@hovsepm all the feedback is addressed and yes this is intended to be merged.

File: build.proj [hunk 5]
```diff
@@ -61,27 +61,27 @@
     <FxTargetList Condition=" '$(Scope)' == 'authentication' ">net45</FxTargetList>
     <ZipExeFolder>$(LibraryToolsFolder)\7-Zip</ZipExeFolder>
     <ZipExe>$(ZipExeFolder)\7z.exe</ZipExe>
+	<NuGetCommand>&quot;$(LibraryToolsFolder)\nuget.exe&quot;</NuGetCommand>
```
Comment (markcowl) [inline build.proj]: fix indentation

File: build.proj [hunk 6]
```diff
@@ -61,27 +61,27 @@
...
+  <ItemGroup Condition="'$(PublishTestProjects)' == 'false' Or '$(PublishTestProjects)' == ''">
     <LibrariesToBuild Include="$(LibrarySourceFolder)\$(Scope)\*.sln" Condition=" '$(Scope)' != 'all'  " />
     <LibrariesToBuild Include="$(LibrarySourceFolder)\**\*.sln" Condition=" '$(Scope)' == 'all'  " />
     <LibraryFxTargetList Include="$(FxTargetList)" />
     <AutoRestLibraryFxTargetList Include="portable;net45" />
-    <ClientRuntimeProjects Include="$(LibrarySourceFolder)\ClientRuntime\**\*.xproj" />
-    <ClientRuntimeTests Include="$(LibrarySourceFolder)\ClientRuntime\*.Tests\*.xproj" />
+    <ClientRuntimeProjects Include="$(LibrarySourceFolder)\ClientRuntime\**\*.xproj"/>
+    <ClientRuntimeTests Include="$(LibrarySourceFolder)\ClientRuntime\*.Tests\*.xproj"/>
+	<ClientRuntimeRootDir Include="%(ClientRuntimeProjects.RootDir)"/>
```
Comment (markcowl) [inline build.proj]: Indentation

File: build.proj [hunk 7]
```diff
@@ -131,16 +131,15 @@
     
     <!--Do not build old libraries except Authentication, which Azure PowerShell still uses-->
     <ItemGroup>
-      <NonAutoRestLibraries Include="$(LibrarySourceFolder)\Authentication\Authentication.sln"
-                            Condition=" '$(Scope)' == 'all' or '$(Scope)' == 'Authentication' " />
+		<NonAutoRestLibraries Include="$(LibrarySourceFolder)\Authentication\Authentication.sln"
```
Comment (markcowl) [inline build.proj]: Indentation

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs (hunk 3 above)
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs]: @shahabhijeet Please address this feedback

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_DataPlane/Class1.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RP1_DataPlane
+{
+    public class Class1
+    {
+        public Class1()
+        {
+        }
```
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_DataPlane/Class1.cs]: Include a method

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_MgmtPlane/Class1.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RP1_MgmtPlane
+{
+    public class Class1
+    {
+        public Class1()
+        {
+        }
```
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_MgmtPlane/Class1.cs]: At least a dummy method for this class.

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP2_Sdk/RP2_SDK.Test/Class1.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RP2_SDK.Test
+{
+    public class Class1
+    {
+        public Class1()
+        {
+        }
```
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP2_Sdk/RP2_SDK.Test/Class1.cs]: same comment

File: tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs [hunk 11]
```diff
@@ -43,6 +53,19 @@ public override bool Execute()
             var netCoreAutoRestLibraries = new List<ITaskItem>();
             var netCoreLibraryTestOnes = new List<ITaskItem>();
             var others =  new List<ITaskItem>();
+            
+            List<string> nPkgsList = null;
+
+            if (NugetPackagesToPublish != null)
```
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs]: Why not simply call String.IsNullOrWhitespace here?

File: tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs [hunk 12]
```diff
@@ -70,7 +94,20 @@ public override bool Execute()
...
-                    nonNetCoreAutoRestLibraries.Add(solution);
+
+                    if (nPkgsList != null)
+                    {
+                        //We need to filter out projects from the final output to build and publish limited set of projects
+                        string projectDirPath = Path.GetDirectoryName(nugetProjects[0]);
+                        string projectDirName = Path.GetFileName(projectDirPath);
+                        string match = nPkgsList.Find((pn) => pn.Equals(projectDirName, System.StringComparison.OrdinalIgnoreCase));
+                        if (!string.IsNullOrEmpty(match))
+                            nonNetCoreAutoRestLibraries.Add(solution);
```
Comment (markcowl) [inline tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs]: Always use brackets for the then clause

**Review by markcowl — CHANGES_REQUESTED** [review]

File: tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs (hunk 11 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs]: thank you, good suggestion, will incorporate that.

File: tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs (hunk 12 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/FilterOutAutoRestLibraries.cs]: done

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs (hunk 3 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectMultiSln/CSProjTestPublish/Class1.cs]: Ok I will add it now :)
I am still not sure, the purpose of adding methods to empty classes. These projects are not meant to be consumed by anyone, they are test assets, whose sole purpose is to get packaged and used in testing publishing functionality.

**Comment by shahabhijeet** [comment]
I will add empty methods :)
But I am not sure the purpose of adding empty methods on test assets whose sole purpose is to get packaged and get used in testing publishing functionality. These test nugget packages will never be consumed by anyone.

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_DataPlane/Class1.cs (hunk 8 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_DataPlane/Class1.cs]: I will add empty methods :)
But I am not sure the purpose of adding empty methods on test assets whose sole purpose is to get packaged and get used in testing publishing functionality. These test nugget packages will never be consumed by anyone.

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_MgmtPlane/Class1.cs (hunk 9 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP1_MgmtPlane/Class1.cs]: done

File: tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP2_Sdk/RP2_SDK.Test/Class1.cs (hunk 10 above)
Comment (shahabhijeet) [inline tools/Microsoft.WindowsAzure.Build.Tasks/Build.Tasks.Tests/TestPublishProjects/MultiProjectSingleSln/RP2_Sdk/RP2_SDK.Test/Class1.cs]: done

File: build.proj (hunk 7 above)
Comment (shahabhijeet) [inline build.proj]: I don't see any indentation issue when I open this file locally.

**Review by markcowl — APPROVED** [review]

**Pull request merged**